pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go

```diff
@@ -0,0 +1,693 @@
+			secretName = accessCred.UserPassword.Source.Secret.SecretName
+		}
+	}
+
+	return secretName
+}
+
+func (l *AccessCredentialManager) reportAccessCredentialResult(vmi *v1.VirtualMachineInstance, succeeded bool, message string) error {
+	l.domainModifyLock.Lock()
+	defer l.domainModifyLock.Unlock()
+
+	domName := api.VMINamespaceKeyFunc(vmi)
+	dom, err := l.virConn.LookupDomainByName(domName)
+	if err != nil {
+		if domainerrors.IsNotFound(err) {
+			return nil
+		} else {
+			log.Log.Reason(err).Error("Getting the domain for completed migration failed.")
+		}
+		return err
+	}
+	defer dom.Free()
+
+	state, _, err := dom.GetState()
+	if err != nil {
+		return err
+	}
+	domainSpec, err := util.GetDomainSpec(state, dom)
+	if err != nil {
+		return err
+	}
+
+	if domainSpec.Metadata.KubeVirt.AccessCredential == nil ||
+		domainSpec.Metadata.KubeVirt.AccessCredential.Succeeded != succeeded ||
+		domainSpec.Metadata.KubeVirt.AccessCredential.Message != message {
+
+		domainSpec.Metadata.KubeVirt.AccessCredential = &api.AccessCredentialMetadata{
+			Succeeded: succeeded,
+			Message:   message,
+		}
+	} else {
+		// nothing to do
+		return nil
+	}
+
+	_, err = util.SetDomainSpecStrWithHooks(l.virConn, vmi, domainSpec)
+	return err
+}
+
+func (l *AccessCredentialManager) watchSecrets(vmi *v1.VirtualMachineInstance) {
+	logger := log.Log.Object(vmi)
+
+	reload := true
+	fileChangeDetected := true
+
+	domName := util.VMINamespaceKeyFunc(vmi)
+
+	// guest agent will force a resync of changes every 'x' minutes
+	forceResyncTicker := time.NewTicker(5 * time.Minute)
+	defer forceResyncTicker.Stop()
+
+	// guest agent will aggregate all changes to secrets and apply them
+	// every 'x' seconds. This could help prevent making multiple qemu
+	// execution calls in the event multiple secret changes are landing within
+	// a small timeframe.
+	handleChangesTicker := time.NewTicker(time.Duration(l.resyncCheckIntervalSeconds) * time.Second)
+	defer handleChangesTicker.Stop()
+
+	for {
+		select {
+		case <-l.watcher.Events:
+			fileChangeDetected = true
+		case err := <-l.watcher.Errors:
+			logger.Reason(err).Errorf("Error encountered while watching downward api secret")
+		case <-forceResyncTicker.C:
+			reload = true
+			logger.Info("Resyncing access credentials due to recurring resync period")
+		case <-handleChangesTicker.C:
+			if fileChangeDetected {
+				reload = true
+				logger.Info("Reloading access credentials because secret changed")
+			}
+		case <-l.stopCh:
+			logger.Info("Signalled to stop watching access credential secrets")
+			return
+		}
+
+		if !reload {
+			continue
+		}
+
+		fileChangeDetected = false
+		reload = false
+		reportedErr := false
+
+		// secret name mapped to authorized_keys in that secret
+		secretMap := make(map[string]string)
+		// filepath mapped to secretNames
+		userSSHMap := make(map[string][]string)
+		// maps users to passwords
+		userPasswordMap := make(map[string]string)
+
+		err := l.pingAgent(domName)
```

Comment:
I wonder if we should distinguish between different types of agent errors and fail early (or alert the user) if the agent is not installed / not available.
Something like we do here: https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-launcher/virtwrap/manager.go#L554

Reply:
take a look at what I'm doing with the ```l.reportAccessCredentialResult``` function in ```pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go```

What happens is whenever an error occurs, that error typically indicates both where it occurred (like during file write) with the libvirt error message appended to the end. That error then gets placed on the domain's metadata which is read by virt-handler. Virt-handler then adds both a condition and reports an event to indicated to the user whether or not access credential syncing succeeded or not (with a message indicating why it didn't succeed)

Maybe i could interpret the libvirt error code, but from what I've seen so far it seems like the libvirt error message itself contains a human readable string that indicates the issue already.

Reply:
Thanks for the explanation! 
Yeah, I think it's good.

What I meant with my comment was that we could stop watching the directories when we determine that the agent is not even installed.
But I think that this won't give the user a chance to install it after seeing the event.